apollo-ios-pagination/Sources/ApolloPagination/GraphQLQueryPager.swift

```diff
@@ -1,342 +1,207 @@
   public func subscribe(onUpdate: @MainActor @escaping (Result<Output, Error>) -> Void) {
-    Task {
-      await pager.subscribe(onUpdate: onUpdate)
-        .store(in: &cancellables)
+    Task { [weak self] in
+      guard let self else { return }
+      let subscription = await self.pager.subscribe(onUpdate: onUpdate)
+      await subscriptions.store(subscription: subscription)
     }
   }
 
-  public var canLoadNext: Bool { canLoadNextSubject.value }
+  /// Whether or not we can load the next page. Initializes with a `false` value that is updated after the initial fetch.
+  public var canLoadNext: Bool = false
+  /// Whether or not we can load the previous page. Initializes with a `false` value that is updated after the initial fetch.
+  public var canLoadPrevious: Bool = false
 
+  /// Reset all pagination state and cancel all in-flight requests.
   public func cancel() {
-    Task {
-      await pager.cancel()
+    Task { [weak self] in
+      guard let self else { return }
+      for completion in await self.completionManager.completions {
+        await completion.execute(error: PaginationError.cancellation)
+      }
+      await self.completionManager.reset()
+      await self.pager.cancel()
     }
   }
 
-  public func loadMore(
+  /// Loads the previous page, if we can.
+  /// - Parameters:
+  ///   - cachePolicy: The Apollo `CachePolicy` to use. Defaults to `fetchIgnoringCacheData`.
+  ///   - completion: An optional error closure that triggers in the event of an error. Defaults to `nil`.
+  public func loadPrevious(
     cachePolicy: CachePolicy = .fetchIgnoringCacheData,
-    completion: (@MainActor () -> Void)? = nil
-  ) throws {
-    Task {
-      try await pager.loadMore(cachePolicy: cachePolicy)
-      await completion?()
+    completion: ((Error?) -> Void)? = nil
+  ) {
+    execute(completion: completion) { [weak self] in
+      try await self?.pager.loadPrevious(cachePolicy: cachePolicy)
+    }
+  }
+
+  /// Loads the next page, if we can.
+  /// - Parameters:
+  ///   - cachePolicy: The Apollo `CachePolicy` to use. Defaults to `fetchIgnoringCacheData`.
+  ///   - completion: An optional error closure that triggers in the event of an error. Defaults to `nil`.
+  public func loadNext(
+    cachePolicy: CachePolicy = .fetchIgnoringCacheData,
+    completion: ((Error?) -> Void)? = nil
+  ) {
+    execute(completion: completion) { [weak self] in
+      try await self?.pager.loadNext(cachePolicy: cachePolicy)
+    }
+  }
+
+  /// Loads all pages.
+  /// - Parameters:
+  ///   - fetchFromInitialPage: Pass true to begin loading from the initial page; otherwise pass false.  Defaults to `true`.  **NOTE**: Loading all pages with this value set to `false` requires that the initial page has already been loaded previously.
+  ///   - completion: An optional error closure that triggers in the event of an error. Defaults to `nil`.
+  public func loadAll(fetchFromInitialPage: Bool = true, completion: ((Error?) -> Void)? = nil) {
```

Comment:
I'm just now thinking... why are `loadNext()` and `loadAll()` using completion blocks instead of being `async throws` functions? If we're using `async/await` anyways, why aren't we using it for the `public` callsites?

Reply:
Oh is the completion ony called if an error occurs? If so, that needs to be more clearly documented.

Is the idea here that you would normally just call `loadAll()` and then move on, not needing to `await` on it because once it gets data back, that will get pushed into the subscribers via the `Combine` subject?

I'm feeling like there is something off with this API design. It's kind of confusing that you get the results by subscribing, but the errors through a completion handler. 

It's unclear to me what the difference is between getting an `Error` from this completion block vs. getting a `Result.failure` with an error from the subscriber's output.

I'm not sure what the right solution to this is yet. Happy to chat about this with you some more though!

Reply:
I see where the confusion lies, and I think you're smelling something real.

The API structure is, from bottom to the top, something like this:

1. `GraphQLQueryPager.Actor`: An `actor` that can operate asynchronously and be easily used in `async` contexts. 
2. `GraphQLQueryPager`: A wrapper around `GraphQLQueryPager.Actor` which allows easier usage in non-async contexts. Think: Existing `UIKit` apps.
3. `AnyGraphQLQueryPager`: A wrapper around `GraphQLQueryPager` which erases its query, allowing the caller to map results to their own model. 

> Oh is the completion ony called if an error occurs? If so, that needs to be more clearly documented.

The completion is always triggered, but provides an `Error` if there was an error of some kind. I suppose it's an alternative to using `Result<Void, Error>`, where the success case just means there's no error.